Message versioning and ROS 2 message translation (#3465)
PX4/PX4-user_guide: en/middleware/uorb.md

The full API is documented in [platforms/common/uORB/uORBManager.hpp](https://github.com/PX4/PX4-Autopilot/blob/main/platforms/common/uORB/uORBManager.hpp).

## Message Versioning

> 

Reviewer: Just FYI, this is how I mark up new content. On release I'll know the version number and replace this. You should also add a release note in https://docs.px4.io/main/en/releases/main.html , ideally with a link to any associated PRs.

I'll review properly when you tell me this is ready and/or when this moves out of draft. At that time would be good to link the various proposals that lead to this in the description.

Glad you are doing this and not me :-)

```suggestion

<Badge type="tip" text="main (PX4 v1.16+)" />

```

Author: @hamishwillee thanks for the early review (: I've applied your suggestions, and with some of the related PRs getting merged I'm coming back to these docs.

There's essentially 2 things to document here:
1. The new versioned messages (already started here)
2. The ROS 2 translation node (to do)

The README in [this PR](https://github.com/PX4/PX4-Autopilot/pull/24113) documents point 2 well, I will now try to capture those points here. Once that's done I will open this PR for review

The full API is documented in [platforms/common/uORB/uORBManager.hpp](https://github.com/PX4/PX4-Autopilot/blob/main/platforms/common/uORB/uORBManager.hpp).

## Message Versioning

<Badge type="tip" text="main (PX4 v1.16+)" />

> Message versioning was introduced to address the challenges of maintaining compatibility across systems where different versions of message definitions may be in use, such as between PX4 and external systems like ROS 2 applications.
> 
> This versioning mechanism supports the [ROS 2 Message Translation Node](../ros2/px4_ros2_msg_translation_node.md), which enables seamless communication between PX4 and ROS 2 applications; when different versions of message definitions are in use, the ROS 2 translation node ensures that messages can be converted and exchanged correctly.

Reviewer: This is a little messy and perhaps can be simplified. In particular
- ROS2 Message Translation node supports versioning, not the versioning supporting the translation node.
 
Perhaps something like:

```suggestion
Message versioning has been in introduced in PX4 v1.16 to make it easier to maintain compatibility between PX4 and ROS 2 versions.
This versioning mechanism uses the [ROS 2 Message Translation Node](../ros2/px4_ros2_msg_translation_node.md) to convert between different versions of messages, allowing seamless communication between PX4 and ROS 2 applications; when different versions of message definitions are in use.
```

Perhaps add a diagram showing where the magic happens?

It would also be good "somewhere here" to mention any forward thinking about the ROS 2 native translation system that they have been talking about for a couple of years. Perhaps as a note at the end.

Author: Good idea, added a diagram and a note concerning ROS native translation.

> ROS2 Message Translation node supports versioning, not the versioning supporting the translation node.

I may be thinking of this wrong, but in my mind message versioning works as a standalone feature (it's just about some messages having a version number incremented, others not). The translation node builds on top of versioned messages. In that sense, the translation supporting the versioning sounds backwards to me. How do you see this?

Reviewer: You are correct. Fortunately this has been fixed by my subedit to my taste.

> Versioned messages are stored in the `msg/versioned/` directory, distinct from their non-versioned counterparts, which reside directly in the `msg/` directory.

Reviewer: Nothing wrong with this, but I'm trying to make the distinction very clear, since I missed this first time.

```suggestion
Versioned messages are stored in the `msg/versioned/` directory (distinct from their non-versioned counterparts, which reside directly in the `msg/` directory).
```

Each versioned message definition includes an additional field: `uint32 MESSAGE_VERSION = X`, where `X` corresponds to the current version of the message.
When a versioned message definition is modified, the version number should be incremented to reflect changes in its structure or semantics.

> Additionally, [message translations](../ros2/px4_ros2_msg_translation_node.md#updating-a-message) must also be updated for versioned messages.

Reviewer: ```suggestion
Additionally, [message translations](../ros2/px4_ros2_msg_translation_node.md#updating-a-message) must also be updated for versioned messages, in order to convert from the old version to the new structure.
```

Versioned messages are designed to remain more stable over time compared to their non-versioned counterparts, as they are intended to be used across multiple releases of PX4 and external systems, ensuring greater compatibility over longer periods.

> For the full list of versioned and non-versioned messages, refer to the full [uORB Message Reference](../msg_docs/index.md).

Reviewer: ```suggestion
For the full list of versioned and non-versioned messages, refer to the [uORB Message Reference](../msg_docs/index.md).
```

For more on PX4 and ROS 2 communication, refer to the page about the [PX4-ROS 2 Bridge](../ros/ros2_comm.md).

## Message/Field Deprecation {#deprecation}

> (TODO: Is this section still relevant? Not seeing any `# DEPRECATED: ...` fields or msgs in latest PX4)

Reviewer: Almost certainly still relevant. We don't want to break Flight Review or other tools. @bkueng will know.

Reviewer 2: Right, it's still relevant but it usually gets forgotten to be added.

The full API is documented in [platforms/common/uORB/uORBManager.hpp](https://github.com/PX4/PX4-Autopilot/blob/main/platforms/common/uORB/uORBManager.hpp).

## Message Versioning

<Badge type="tip" text="main (PX4 v1.16+)" />

Message versioning has been in introduced in PX4 v1.16 (main) to make it easier to maintain compatibility between PX4 and ROS 2 versions.

Versioned messages include an additional field `uint32 MESSAGE_VERSION = x`, where `x` corresponds to the current version of the message.

> Versioned message files are stored separetly from their non-versioned counterpart.

Reviewer 3: ```suggestion
Versioned message files are stored separately from their non-versioned counterpart.
```

Non-versioned messages remain in the `msg/` and `srv/` directories respectively.

This versioning mechanism supports the [ROS 2 Message Translation Node](../ros2/px4_ros2_msg_translation_node.md), which enables seamless communication between PX4 and ROS 2 applications; when different versions of message definitions are in use, the ROS 2 translation node ensures that messages can be converted and exchanged correctly.

Versioned messages are designed to remain more stable over time compared to their non-versioned counterparts, as they are intended to be used across multiple releases of PX4 and external systems, ensuring greater compatibility over longer periods.

Updating a versioned message involves more steps compared to updating a non-versioned one.
Please see the section on [Updating a Versioned Message](../ros2/px4_ros2_msg_translation_node.md#updating-a-versioned-message).

For the full list of versioned and non-versioned messages, refer to the [uORB Message Reference](../msg_docs/index.md).

For more on PX4 and ROS 2 communication, refer to the page about the [PX4-ROS 2 Bridge](../ros/ros2_comm.md).

::: note

> ROS 2 plans to natively support message versioning in the future, but this is not implmemented yet.

Reviewer 3: ```suggestion
ROS 2 plans to natively support message versioning in the future, but this is not implememented yet.
```

The full API is documented in [platforms/common/uORB/uORBManager.hpp](https://github.com/PX4/PX4-Autopilot/blob/main/platforms/common/uORB/uORBManager.hpp).

## Message Versioning

<Badge type="tip" text="main (PX4 v1.16+)" />

Optional message versioning was introduced in PX4 v1.16 (main) to make it easier to maintain compatibility between PX4 and ROS 2 versions.
Versioned messages are designed to remain more stable over time compared to their non-versioned counterparts, as they are intended to be used across multiple releases of PX4 and external systems, ensuring greater compatibility over longer periods.

Versioned messages include an additional field `uint32 MESSAGE_VERSION = x`, where `x` corresponds to the current version of the message.

Versioned and non-versioned messages are separated in the file system:

> - Non-versioned topic message files and service message files remain in the [`msg/`](https://github.com/PX4/PX4-Autopilot/tree/main/msg) and [`srv/`](https://github.com/PX4/PX4-Autopilot/tree/main/srv) directories, respectively.

Reviewer: I've subedited this a little, mainly "to my taste".

It is still a problem for me that this is the first and only mention of "topic message files and service message files" and that there is no definition of what these are for, or where they are from. If you're from ROS-land perhaps you know, but I can assure you that many readers will not.

Can you create a section in this topic which outlines the differences/similarities, and when you should create one or the other. Can be a different PR.
